### ISSUE TITLE
DataCollection Cycle Iteration

### ISSUE THREAD

**Issue by termi-official** [mfem/mfem#2574]
Hi folks,

what is the intended way to obtain cycle information for loadable ` DataCollections ` ?
 
To elaborate, let us take the example of converting the full collection in format A to a full collection in format B. Then, looking at the convert-dc miniapp, an external script looping over the app must be utilized. Further this script must know all cycles (in the case that the cycle is not a "continuous" integer range) beforehand. This is especially problematic for more complex post-processing steps like computing the space-time distance between two solutions.

**Comment by v-dobrev** [comment]
To a large extend, how different cycles are stored as files depends on the specific `DataCollection` and the visualization tool they are to be used with. For example, you can take a look at the outputs from these two runs of Example 9p:
```console
mpirun -np 4 ./ex9p -visit
mpirun -np 4 ./ex9p -paraview
```
Does this answer your question?

**Comment by termi-official** [comment]
Thanks for the quick answer. I was probably not specific enough with my question. That the "metafiles" containing the cycle information is different is part of my problem. I try to develop some post-processing tools in MFEM, e.g. computing errors of simulations (also done with MFEM) w.r.t. computed reference solutions. I started developing a custom ` DataCollection ` whose only purpose is to store a CSV formatted file containing the relation between the cycle index and time points. However revisiting for example the ` VisitDataCollection ` it should also be possible to extract the same information from the root file.

Basically I am searching for methods like ` std::vector<int> cycles = dc->GetAllCycles() `. But from your answer I may take away that  ` DataCollections ` are really just ment to be containers for visualizations?

**Comment by v-dobrev** [comment]
`DataCollection`s can be used for other purposed, not just visualization -- for example, for restarting a time-dependent simulation.

Typically, `DataCollection` stores the data for only one cycle, so a method like `dc->GetAllCycles()` does not necessarily make sense. Maybe for some sub-classes it makes sense -- it can try to access that information from the file system based on the output scheme used by the particular sub-class. Is that what you are thinking?

**Comment by termi-official** [comment]
Exactly. I often require to reload either data at specific time points for post processing or even interpolate in time (by loading pairs of solutions at $t_{n}$ and $t_{n-1}$). I already have some subclass `MFEMDataCollection` which can do this, but it is rather fragile, hence my question if I overlooked something more stable in the current MFEM version.

**Comment by v-dobrev** [comment]
I think adding such virtual method in the base class will be useful. Do you want to propose a PR based on what you have? We can add implementations for the different `DataCollection`s as we need them -- the ones that don't support this method can just error out.

**Comment by termi-official** [comment]
Sounds good. I will open a PR next week or so - however the current implementation which I have is rather fragile and probably does not work in general for all kinds of simulations.

**Comment by termi-official** [comment]
Bump. Still need some time, sorry.

**Comment by tzanio** [comment]
Discussion continued in #2689
